lib/r/result.rb

```diff
@@ -597,6 +600,22 @@ def self.ok(value)
     Ok.new(value)
   end
 
+  sig do
+    type_parameters(:Ok, :Err)
+      .params(
+        blk: T.proc.returns(Result[T.type_parameter(:Ok), T.type_parameter(:Err)]),
+      )
+      .returns(Result[T.type_parameter(:Ok), T.type_parameter(:Err)])
+  end
+  def self.propagate!(&blk)
+    # TODO: using singleton class instance variables is obviously bad and not
+    # thread-safe but demonstrates how it might work.
+    @ball = T.let(Object.new, Object)
```

Comment:
Though most examples online of `Kernel.catch` use symbols, it actually uses `object_id` to catch a thrown object. The block you pass to `catch` gets passed a unique object that is specifically intended to be `throw`n for this purpose, but it seemed imprudent to expose that object to the block that the consumer's of this library see, so instead this just creates a new `Object` so that it will only catch this specific object's `object_id` and consumer code can't accidentally early-return if they use `throw` inside the block for some reason.